### ISSUE TITLE
Store data on integration points

### ISSUE THREAD

**Issue by mrp089** [SimVascular/svFSIplus#179]
### Use Case

All constrained-mixture-based (CMM) growth and remodeling (G&R) solid material models need to store internal data on each integration (Gauss) point. Different CMM materials store different sets of information:
- full CMM (#2): every past time step
- homogenized CMM (#33): previous time step
- equilibrated CMM (#49): pre-loading time step

### Problem

We currently store all global data on a node/dof-basis, which doesn't fit the G&R models.

### Solution

There are two ways to implement this:
1. Store a `n`-sized vector on each integration point, where `n` is the number if internal variables
2. Store an `n x elem`-sized matrix globally, where `elem` is the total number of elements

I think approach 1 might be easier to implement with parallelization in mind. The elements are already distributed to processors and the data is only needed within each element. An example of this would be the [FEMaterialPoint](https://github.com/febiosoftware/FEBio/blob/develop/FECore/FEMaterialPoint.cpp) in `FEBio`, see usage [here](https://github.com/mrp089/FEBio_CMM_plugin/blob/3b3ab7991f795cad219da0fdcd0ea703be5b4827/FEMbeCmm.cpp#L30-L44). 

### Alternatives considered

We discussed with @MatteoSalvador storing all data in a node-based format using back-and-forth projection using the elements' shape functions. The advantage of this would be that we get a regularization of G&R variables by the shape functions. Also, it would fit the FEM way of storing all data on a nodal basis.

After thinking about it more, I think we can't do this easily. These internal variables don't appear (directly) in the weak form and are not part of the unknowns. The projection would make everything we compute non-local to the elements and complicate the linearization. Furthermore, we would need a way of spatially interpolating tensors. Happy to discuss more!

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct and Contributing Guidelines

**Comment by ktbolt** [comment]
This G&R internal data is a state variable yes? State variables (velocity, displacement, etc.) are stored as global arrays in the `ComMod` class so if we want to follow how things are currently implemented then G&R internal data should be stored as an `n x elem-sized matrix` in `ComMod`.

On the other hand, we may want to break with tradition and create some classes and some sort of interface for this, just need to get it to work with the parallel distribution code and also with the post processing code (averaging element data to shared nodes).

**Comment by ktbolt** [comment]
@mrp089 There has been no update here for a year. Is this still being worked on ? There is no one even assigned to work on this.

**Comment by mrp089** [comment]
I don't know of anybody working on this. However, this will still be required once 3D constrained-mixture-based G&R models are implemented.

**Comment by alisonmarsden** [comment]
I would suggest we develop a system for categorizing 1) issues that are actively being worked on and are assigned to someone and 2) future plans that we want to keep track of but are not actively being developed now.   Suggestions for how to track these differently?


From: Martin R. Pfaller ***@***.***>
Date: Thursday, February 6, 2025 at 8:05 AM
To: SimVascular/svMultiPhysics ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [SimVascular/svMultiPhysics] Store data on integration points (Issue #179)

I don't know of anybody working on this. However, this will still be required once 3D constrained-mixture-based G&R models are implemented.

**Comment by ktbolt** [comment]
@alisonmarsden I was thinking the same thing! 

I know if an Issue I opened is for future long-term development but I don't know about Issues opened by other team members.

Here are a couple of options I can think of

1) Create a custom `Future` GitHub [Label](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) to identify long-term future development Issues.

2) Use GitHub [Projects](https://docs.github.com/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects) to manage Issues. This would require some effort to use and might be overkill since out team is small with typically a single developer is working on an Issue.

I think 1) would be easy and allow us to immediately see which Issues are long-term. When we actually start work on it we can change the label. 

@mrp089 What do you think?

**Comment by alisonmarsden** [comment]
I agree with both of these ideas.   The first should be easy and we can do right away.    Using Projects would be a good as well, but may take a bit more time and thought.    We could discuss this in the next meeting.


From: Dave Parker ***@***.***>
Date: Thursday, February 6, 2025 at 11:56 AM
To: SimVascular/svMultiPhysics ***@***.***>
Cc: Alison Lesley Marsden ***@***.***>, Mention ***@***.***>
Subject: Re: [SimVascular/svMultiPhysics] Store data on integration points (Issue #179)

@alisonmarsden<https://github.com/alisonmarsden> I was thinking the same thing!

I know if an Issue I opened is for future long-term development but I don't know about Issues opened by other team members.

Here are a couple of options I can think of

  1.  Create a custom Future GitHub Label<https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels> to identify long-term future development Issues.
  2.  Use GitHub Projects<https://docs.github.com/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects> to manage Issues. This would require some effort to use and might be overkill since out team is small with typically a single developer is working on an Issue.

I think 1) would be easy and allow us to immediately see which Issues are long-term. When we actually start work on it we can change the label.

@mrp089<https://github.com/mrp089> What do you think?
